File: processor/processor/clip/Clip.py
from datetime import datetime
from enum import Enum
from typing import Optional

class ThemeShadow(str, Enum):
    NONE = "None"
    SMALL = "Small"
    MEDIUM = "Medium"
    LARGE = "Large"

class ThemeStroke(str, Enum):
    NONE = "None"
    SMALL = "Small"
    MEDIUM = "Medium"
    LARGE = "Large"

class ThemeEmojiPosition(str, Enum):
    TOP = "Top"
    BOTTOM = "Bottom"

class Range:
    def __init__(self, start: int, end: int):
        self.start: int = start
        self.end: int = end

    def __repr__(self):
        return f"Range(start={self.start!r}, end={self.end!r})"


class Fragment:
    def __init__(self, x: int, y: int, width: int, height: int):
        self.x: int = x
        self.y: int = y
        self.width: int = width
        self.height: int = height

    def __repr__(self):
        return f"Fragment(x={self.x!r}, y={self.y!r}, width={self.width!r}, height={self.height!r})"


class Section:
    def __init__(
        self, order: int, start: int, end: int, display: str, fragments: list[Fragment]
    ):
        self.order: int = order
        self.start: int = start
        self.end: int = end
        self.display: str = display
        self.fragments: list[Fragment] = fragments

    def __repr__(self):
        return f"Section(order={self.order!r}, start={self.start!r}, end={self.end!r}, display={self.display!r}, fragments={self.fragments!r})"


class Theme:
    def __init__(
        self,
        themeFont: str,
        themeFontColor: str,
        themeSize: int,
        themePosition: int,
        themeMainColor: str,
        themeSecondaryColor: str,
        themeThirdColor: str,
        themeStroke: ThemeStroke,
        themeStrokeColor: str,
        themeShadow: ThemeShadow,
        themeUpperText: bool,
        themeEmoji: bool,
        themeEmojiPosition: ThemeEmojiPosition,
    ):
        self.themeFont: str = themeFont
        self.themeFontColor: str = themeFontColor
        self.themeSize: int = themeSize
        self.themePosition: int = themePosition
        self.themeMainColor: str = themeMainColor
        self.themeSecondaryColor: str = themeSecondaryColor
        self.themeThirdColor: str = themeThirdColor
        self.themeStroke: ThemeStroke = themeStroke
        self.themeStrokeColor: str = themeStrokeColor
        self.themeShadow: ThemeShadow = themeShadow
        self.themeUpperText: bool = themeUpperText
        self.themeEmoji: bool = themeEmoji
        self.themeEmojiPosition: ThemeEmojiPosition = themeEmojiPosition

    def __repr__(self):
        return (
            "Theme(" + ", ".join([f"{k}={v!r}" for k, v in self.__dict__.items()]) + ")"
        )


class Clip:
    def __init__(
        self,
        id: str,
        sourceId: str,
        name: str,
        url: Optional[str],
        processing: bool,
        width: int,
        height: int,
        createdAt: datetime,
        updatedAt: datetime,
        range: Range,
        sections: list[Section],
        theme: Theme,
    ):
        self.id: str = id
        self.sourceId: str = sourceId
        self.name: str = name
        self.url: Optional[str] = url
        self.processing: bool = processing
        self.width: int = width
        self.height: int = height
        self.createdAt: datetime = createdAt
        self.updatedAt: datetime = updatedAt

        self.range: Range = range
        self.sections: list[Section] = sections

        self.theme: Theme = theme

    def __repr__(self):
        return f"Clip(id={self.id!r}, sourceId={self.sourceId!r}, name={self.name!r}, url={self.url!r}, processing={self.processing!r}, width={self.width!r}, height={self.height!r}, createdAt={self.createdAt!r}, updatedAt={self.updatedAt!r}, range={self.range!r}, sections={self.sections!r}), theme={self.theme!r}"

File: application/src/server/entities/clip/domain/Clip.ts
import { newDate } from "@/utils/newDate";
import { v4 as uuidv4 } from "uuid";
import z from "zod";

export enum ThemeFont {
  Komika = "Komika",
}

export enum ThemeShadow {
  None = "None",
  Small = "Small",
  Medium = "Medium",
  Large = "Large",
}

export enum ThemeStroke {
  None = "None",
  Small = "Small",
  Medium = "Medium",
  Large = "Large",
}

export enum ThemeEmojiPosition {
  Top = "Top",
  Bottom = "Bottom",
}

export enum DisplayName {
  One = "One",
  Column = "Column",
  Row = "Row",
}

export const SectionSchema = z.object({
  start: z.number(),
  end: z.number(),
  display: z.nativeEnum(DisplayName),
  fragments: z.array(
    z.object({
      order: z.number(),
      x: z.number(),
      y: z.number(),
      size: z.number(),
    }),
  ),
});
export type SectionType = z.infer<typeof SectionSchema>;

export const ThemeSchema = z.object({
  themeFont: z.nativeEnum(ThemeFont),
  themeFontColor: z.string(),
  themeSize: z.number(),
  themeMainColor: z.string(),
  themeSecondaryColor: z.string(),
  themeThirdColor: z.string(),
  themeShadow: z.nativeEnum(ThemeShadow),
  themeStroke: z.nativeEnum(ThemeStroke),
  themeStrokeColor: z.string(),
  themeUpperText: z.boolean(),
  themePosition: z.number(),
  themeEmoji: z.boolean(),
  themeEmojiPosition: z.nativeEnum(ThemeEmojiPosition),
});

export type ThemeType = z.infer<typeof ThemeSchema>;

export const ClipSchema = z.object({
  id: z.string(),
  name: z.string(),
  processing: z.boolean().optional(),
  sourceId: z.string(),
  range: z.object({
    start: z.number(),
    end: z.number(),
  }),
  width: z.number(),
  height: z.number(),
  sections: z.array(SectionSchema),
  theme: ThemeSchema,
});

export type ClipType = z.infer<typeof ClipSchema>;

export const defaultWidth = 0;
export const defaultHeight = 0;

export const defaultDisplay = DisplayName.One;
export const defaultFragments = [
  {
    order: 0,
    x: 0,
    y: 0,
    size: 1/2,
  },
];

export const defaultTheme: ThemeType = {
  themeFont: ThemeFont.Komika,
  themeFontColor: "#FFFFFF",
  themeSize: 12,
  themePosition: 50,
  themeMainColor: "#d4c591",
  themeSecondaryColor: "#63edc3",
  themeThirdColor: "#9560c6",
  themeShadow: ThemeShadow.None,
  themeStroke: ThemeStroke.Small,
  themeStrokeColor: "#000000",
  themeUpperText: false,
  themeEmoji: false,
  themeEmojiPosition: ThemeEmojiPosition.Top,
};

export class Clip {
  id: string;
  companyId: string;
  sourceId: string;
  name: string;
  processing: boolean;
  range: {
    start: number;
    end: number;
  };
  width: number;
  height: number;
  sections: SectionType[];
  theme: ThemeType;
  createdAt: Date;
  updatedAt: Date;

  constructor({
    id,
    companyId,
    name,
    processing,
    sourceId,
    range,
    width,
    height,
    sections,
    theme,
    createdAt,
    updatedAt,
  }: {
    id: string;
    companyId: string;
    name: string;
    processing: boolean;
    sourceId: string;
    range: {
      start: number;
      end: number;
    };
    width: number;
    height: number;
    sections: SectionType[];
    theme: ThemeType;
    createdAt: Date;
    updatedAt: Date;
  }) {
    this.id = id;
    this.companyId = companyId;
    this.name = name;
    this.processing = processing;
    this.sourceId = sourceId;
    this.range = range;
    this.width = width;
    this.height = height;
    this.sections = sections;
    this.theme = theme;
    this.createdAt = createdAt;
    this.updatedAt = updatedAt;
  }

  static new({
    sourceId,
    companyId,
    start,
    end
  }: {
    sourceId: string;
    companyId: string;
    start: number;
    end: number;
  }) {
    return new Clip({
      id: uuidv4(),
      companyId: companyId,
      name: "New clip",
      sourceId: sourceId,
      range: {
        start: start,
        end: end,
      },
      processing: false,
      width: defaultWidth,
      height: defaultHeight,
      sections: [
        {
          start: 0,
          end: end - start,
          display: defaultDisplay,
          fragments: defaultFragments,
        },
      ],
      theme: defaultTheme,
      createdAt: newDate(),
      updatedAt: newDate(),
    });
  }

  update(input: ClipType) {
    this.name = input.name;
    this.range = input.range;
    this.width = input.width;
    this.height = input.height;
    this.sections = input.sections;
    this.theme = input.theme;
    this.updatedAt = newDate();
  }

  toPrimitives() {
    return {
      id: this.id,
      companyId: this.companyId,
      name: this.name,
      processing: this.processing,
      sourceId: this.sourceId,
      range: this.range,
      width: this.width,
      height: this.height,
      sections: this.sections,
      theme: this.theme,
      createdAt: this.createdAt,
      updatedAt: this.updatedAt,
    };
  }

  finishProcessing() {
    this.processing = false;
    this.updatedAt = newDate();
  }
}
